File: core/src/main/scala/eu/ostrzyciel/jelly/core/ProtoEncoder.scala
package eu.ostrzyciel.jelly.core

import eu.ostrzyciel.jelly.core.internal.*
import eu.ostrzyciel.jelly.core.proto.v1.*

import scala.collection.mutable

object ProtoEncoder:
  /**
   * Parameters passed to the Jelly encoder.
   *
   * New fields may be added in the future, but always with a default value and in a sequential order.
   * However, it is still recommended to use named arguments when creating this object.
   *
   * @param options options for this stream (required)
   * @param enableNamespaceDeclarations whether to allow namespace declarations in the stream.
   *                                    If true, this will raise the stream version to 2 (Jelly 1.1.0). Otherwise,
   *                                    the stream version will be 1 (Jelly 1.0.0).
   * @param maybeRowBuffer              optional buffer for storing stream rows that should go into a stream frame.
   *                                    If provided, the encoder will append the rows to this buffer instead of
   *                                    returning them, so methods like `addTripleStatement` will return Seq().
   */
  final case class Params(
    options: RdfStreamOptions,
    enableNamespaceDeclarations: Boolean = false,
    maybeRowBuffer: Option[mutable.Buffer[RdfStreamRow]] = None,
  )

/**
 * Base trait for RDF stream encoders.
 * @tparam TNode type of RDF nodes in the library
 * @tparam TTriple type of triple statements in the library
 * @tparam TQuad type of quad statements in the library
 * @tparam TQuoted Unused since 2.7.0. Can be set to `Nothing` or `?`.
 */
trait ProtoEncoder[TNode, -TTriple, -TQuad, -TQuoted]
  extends ProtoEncoderBase[TNode, TTriple, TQuad] with RowBufferAppender:

  /**
   * RdfStreamOptions for this encoder.
   */
  val options: RdfStreamOptions

  /**
   * Whether namespace declarations are enabled for this encoder.
   */
  val enableNamespaceDeclarations: Boolean

  /**
   * Buffer for storing stream rows that should go into a stream frame.
   */
  val maybeRowBuffer: Option[mutable.Buffer[RdfStreamRow]]

  /**
   * Add an RDF triple statement to the stream.
   *
   * If your library does not support quad objects, use `addTripleStatement(s, p, o)` instead.
   *
   * @param triple triple to add
   * @return iterable of stream rows
   * @throws RdfProtoSerializationError if the library does not support triple objects or
   *                                    if a serialization error occurs.
   */
  final def addTripleStatement(triple: TTriple): Iterable[RdfStreamRow] =
    addTripleStatement(
      converter.getTstS(triple),
      converter.getTstP(triple),
      converter.getTstO(triple)
    )

  /**
   * Add an RDF triple statement to the stream.
   * @param subject subject
   * @param predicate predicate
   * @param `object` object
   * @return iterable of stream rows
   * @since 2.9.0
   * @throws RdfProtoSerializationError if a serialization error occurs
   */
  def addTripleStatement(subject: TNode, predicate: TNode, `object`: TNode): Iterable[RdfStreamRow]

  /**
   * Add an RDF quad statement to the stream.
   *
   * If your library does not support quad objects, use `addQuadStatement(s, p, o, g)` instead.
   *
   * @param quad quad to add
   * @return iterable of stream rows
   * @throws RdfProtoSerializationError if the library does not support quad objects or
   *                                    if a serialization error occurs.
   */
  final def addQuadStatement(quad: TQuad): Iterable[RdfStreamRow] =
    addQuadStatement(
      converter.getQstS(quad),
      converter.getQstP(quad),
      converter.getQstO(quad),
      converter.getQstG(quad)
    )

  /**
   * Add an RDF quad statement to the stream.
   *
   * @param subject subject
   * @param predicate predicate
   * @param `object` object
   * @param graph graph
   * @return iterable of stream rows
   * @since 2.9.0
   * @throws RdfProtoSerializationError if a serialization error occurs
   */
  def addQuadStatement(subject: TNode, predicate: TNode, `object`: TNode, graph: TNode): Iterable[RdfStreamRow]

  /**
   * Signal the start of a new (named) delimited graph in a GRAPHS stream.
   * Null value is interpreted as the default graph.
   *
   * @param graph graph node
   * @return iterable of stream rows
   * @throws RdfProtoSerializationError if a serialization error occurs
   */
  def startGraph(graph: TNode): Iterable[RdfStreamRow]

  /**
   * Signal the start of the default delimited graph in a GRAPHS stream.
   *
   * @return iterable of stream rows
   * @throws RdfProtoSerializationError if a serialization error occurs
   */
  def startDefaultGraph(): Iterable[RdfStreamRow]

  /**
   * Signal the end of a delimited graph in a GRAPHS stream.
   *
   * @return iterable of stream rows (always of length 1)
   * @throws RdfProtoSerializationError if a serialization error occurs
   */
  def endGraph(): Iterable[RdfStreamRow]

  /**
   * Declare a namespace in the stream.
   * This is equivalent to the PREFIX directive in Turtle.
   *
   * @param name     short name of the namespace (without the colon)
   * @param iriValue IRI of the namespace
   * @return iterable of stream rows
   * @throws RdfProtoSerializationError if a serialization error occurs
   */
  def declareNamespace(name: String, iriValue: String): Iterable[RdfStreamRow]

File: core-java/src/main/java/eu/neverblink/jelly/core/ProtoEncoder.java
package eu.neverblink.jelly.core;

import eu.neverblink.jelly.core.memory.RowBuffer;
import eu.neverblink.jelly.core.internal.EncoderBase;
import eu.neverblink.jelly.core.proto.v1.RdfQuad;
import eu.neverblink.jelly.core.proto.v1.RdfStreamOptions;
import eu.neverblink.jelly.core.proto.v1.RdfTriple;

/**
 * Base interface for RDF stream encoders.
 * @param <TNode> type of RDF nodes in the library
 */
public abstract class ProtoEncoder<TNode>
    extends EncoderBase<TNode>
    implements RdfBufferAppender<TNode>, RdfHandler.AnyRdfHandler<TNode> {

    /**
     * Parameters passed to the Jelly encoder.
     * <p>
     * New fields may be added in the future, but always with a default value and in a sequential order.
     * WARNING: PLEASE USE .of TO CREATE NEW INSTANCES, otherwise your code will break when new fields are added.
     *
     * @param options options for this stream (required)
     * @param enableNamespaceDeclarations whether to allow namespace declarations in the stream.
     *      If true, this will raise the stream version to 2 (Jelly 1.1.0). Otherwise,
     *      the stream version will be 1 (Jelly 1.0.0).
     * @param rowBuffer buffer for storing stream rows that should go into a stream frame.
     *      The encoder will append the rows to this buffer.
     */
    public record Params(RdfStreamOptions options, boolean enableNamespaceDeclarations, RowBuffer rowBuffer) {
        /**
         * Creates a new instance of Params.
         * @param options options for this stream (required)
         * @param enableNamespaceDeclarations whether to allow namespace declarations in the stream.
         * @param rowBuffer buffer for storing stream rows that should go into a stream frame.
         * @return a new instance of Params
         */
        public static Params of(RdfStreamOptions options, boolean enableNamespaceDeclarations, RowBuffer rowBuffer) {
            return new Params(options, enableNamespaceDeclarations, rowBuffer);
        }

        public Params withOptions(RdfStreamOptions options) {
            return new Params(options, enableNamespaceDeclarations, rowBuffer);
        }

        public Params withEnableNamespaceDeclarations(boolean enableNamespaceDeclarations) {
            return new Params(options, enableNamespaceDeclarations, rowBuffer);
        }

        public Params withRowBuffer(RowBuffer rowBuffer) {
            return new Params(options, enableNamespaceDeclarations, rowBuffer);
        }
    }

    /**
     * RdfStreamOptions for this encoder.
     */
    protected final RdfStreamOptions options;

    /**
     * Whether namespace declarations are enabled for this encoder.
     */
    protected final boolean enableNamespaceDeclarations;

    /**
     * Buffer for storing stream rows that should go into a stream frame.
     */
    protected final RowBuffer rowBuffer;

    protected ProtoEncoder(ProtoEncoderConverter<TNode> converter, Params params) {
        super(converter);
        this.options = params.options
            .clone()
            // Override whatever the user set in the options.
            // If namespace declarations are enabled, we need to use Jelly 1.1.x.
            .setVersion(
                params.enableNamespaceDeclarations
                    ? JellyConstants.PROTO_VERSION_1_1_X
                    : JellyConstants.PROTO_VERSION_1_0_X
            );
        this.enableNamespaceDeclarations = params.enableNamespaceDeclarations;
        this.rowBuffer = params.rowBuffer;
    }

    @Override
    protected int getNameTableSize() {
        return options.getMaxNameTableSize();
    }

    @Override
    protected int getPrefixTableSize() {
        return options.getMaxPrefixTableSize();
    }

    @Override
    protected int getDatatypeTableSize() {
        return options.getMaxDatatypeTableSize();
    }
    
    @Override
    protected RdfTriple.Mutable newTriple() {
        return rowBuffer.newTriple();
    }
    
    @Override
    protected RdfQuad.Mutable newQuad() {
        return rowBuffer.newQuad();
    }

    /**
     * Returns the options for this encoder.
     * @return the options for this encoder
     */
    public RdfStreamOptions getOptions() {
        return options;
    }

    /**
     * Returns the internal row buffer.
     * @return row buffer
     */
    public RowBuffer getRowBuffer() {
        return rowBuffer;
    }
}
